In [1]:
pip install pandas

In [3]:
import pandas as pd

In [5]:
df=pd.read_csv("Downloads/edtech_student_learning_data.csv")

In [7]:
#1.Group the data by course and calculate the average attendance percentage.
df.groupby("course")["attendance_pct"].mean()

course
AI & ML           72.983095
Data Analytics    73.146713
Data Science      73.632578
Generative AI     74.035648
Name: attendance_pct, dtype: float64

interpretation of average attendance percentage is Generative AI like 74% has the best average attendance and AI & ML has like 72%  the lowest attendance

In [10]:
#2. Group the data by city and count the number of students in each city.
df.groupby("city")["student_id"].count()

city
Bangalore     82
Chennai       73
Delhi         68
Hyderabad    100
Mumbai        95
Pune          87
Name: student_id, dtype: int64

interpretation of it show the total number of student enrolled from each city the hydrabad having the highest count and delhi the lowest count

In [13]:
#3. Group the data by batch_type and calculate the average quiz score.
df.groupby("batch_type")["avg_quiz_score"].mean()

batch_type
Weekday    69.628099
Weekend    69.577947
Name: avg_quiz_score, dtype: float64

interpretation of the weekday batches have an average quiz score of 69.62 is this highest and lowest quiz score is an avarage are 69.58

In [16]:
#4. Group the data by placement_status and calculate the average total_score.
df.groupby("placement_status")["avg_quiz_score"].mean()

placement_status
Not Placed    67.662722
Placed        70.577381
Name: avg_quiz_score, dtype: float64

interpretation of placement total_score are student who got placed have a slightly higher like 70.57 average quiz score the those not placed is that comparatively lowest 67.66

In [19]:
#5. Group the data by course and city and calculate the number of students in each combination.
df.groupby(["course", "city"])["student_id"].count()

course          city     
AI & ML         Bangalore    21
                Chennai      14
                Delhi        16
                Hyderabad    19
                Mumbai       30
                Pune         26
Data Analytics  Bangalore    25
                Chennai      18
                Delhi        21
                Hyderabad    33
                Mumbai       26
                Pune         20
Data Science    Bangalore    19
                Chennai      24
                Delhi        12
                Hyderabad    31
                Mumbai       20
                Pune         22
Generative AI   Bangalore    17
                Chennai      17
                Delhi        19
                Hyderabad    17
                Mumbai       19
                Pune         19
Name: student_id, dtype: int64

interpretation of shows how many students are enrolled in each course in each city the student count per course city pair

In [22]:
#6. Group the data by course and calculate the maximum, minimum, and average project score.
df.groupby("course") ["project_score"].agg(["max","min","mean"])

max  min       mean
course                             
AI & ML          99   50  74.158730
Data Analytics   99   50  75.839161
Data Science     99   50  75.992188
Generative AI    98   50  75.037037

in this interpritation are it groups a dataframe df by the “course” column and then calculates three statistics for the “project_score” in each course:
 max – the highest project score, min – the lowest project score,mean – the average project score.also seen in this table.

In [25]:
#7. Group the data by city and calculate total and average learning hours.
df.groupby("city")["learning_hours"].agg(["sum","mean"])

sum       mean
city                      
Bangalore  1870  22.804878
Chennai    1557  21.328767
Delhi      1506  22.147059
Hyderabad  2206  22.060000
Mumbai     2150  22.631579
Pune       1886  21.678161

in this interpritation are the highest learning hours in city mumbai and the lowest learning hours in city are delhi

In [28]:
#8. Group the data by course and placement_status and count students.
df.groupby(["course", "placement_status"])["student_id"].count()

course          placement_status
AI & ML         Not Placed          48
                Placed              78
Data Analytics  Not Placed          44
                Placed              99
Data Science    Not Placed          40
                Placed              88
Generative AI   Not Placed          37
                Placed              71
Name: student_id, dtype: int64

interpritation are the highest score are data analysis in 99 and the lowest placement are generative ai in 37 

In [31]:
#9. Filter grouped results to display only courses with average attendance percentage greater than 80.
avg_attendance=df.groupby("course")["attendance_pct"].mean()
avg_attendance[avg_attendance>80]

Series([], Name: attendance_pct, dtype: float64)

interpretation are in the code there is no average attendance greater than 80 thats why no show the list

In [34]:
#10. Filter grouped results to show only cities with more than 50 students.
avg_students=df.groupby("city")["student_id"].count()
avg_students[avg_students>50]

city
Bangalore     82
Chennai       73
Delhi         68
Hyderabad    100
Mumbai        95
Pune          87
Name: student_id, dtype: int64

interpretation are in the code are the table lists the cities with more than 50 students and their highest student cound city is hydrabad 100 and the lowest student count city are Bangalore82 

In [37]:
#11. Perform multiple aggregations on course to calculate average quiz score, average assignment score, and average project score.
df.groupby("course")[["avg_quiz_score","assignment_score","project_score"]].agg("mean")

avg_quiz_score  assignment_score  project_score
course                                                         
AI & ML              69.023810         71.642857      74.158730
Data Analytics       69.041958         71.545455      75.839161
Data Science         69.960938         72.656250      75.992188
Generative AI        70.592593         72.166667      75.037037

The code performs multiple aggregations on the “course” column to compute three average scores are show in the four column like Average quiz score – the mean score students got in quizzes for each course. Average assignment score – the mean score students achieved on assignments per course.Average project score – the mean score students earned on projects for each course.

In [40]:
#12. Perform named aggregations on course to calculate average_attendance, max_project_score, and mean_total_score.
df["total_score"] = df["avg_quiz_score"] + df["project_score"]
result = df.groupby("course").agg(
    average_attendance=("attendance_pct", "mean"),
    max_project_score=("project_score", "max"),
    mean_total_score=("total_score", "mean"))
df.head()

student_id student_name  gender       city enrollment_date         course  \
0        1001    Student_1    Male  Hyderabad      2024-11-25  Generative AI   
1        1002    Student_2  Female  Hyderabad      2024-09-29        AI & ML   
2        1003    Student_3    Male       Pune      2024-12-05   Data Science   
3        1004    Student_4    Male      Delhi      2024-03-09        AI & ML   
4        1005    Student_5    Male     Mumbai      2024-02-03  Generative AI   

  batch_type  total_sessions  sessions_attended  avg_quiz_score  \
0    Weekday             120                108              59   
1    Weekday             120                110              57   
2    Weekday             120                 73              49   
3    Weekend             120                 71              64   
4    Weekday             120                101              41   

   assignment_score  project_score  final_rating  learning_hours  \
0                96             66           3.7               6   
1                93             83           4.5              33   
2                72             73           3.6               8   
3                91             81           2.7               5   
4                55             50           2.1              23   

  placement_status  feedback_score  attendance_pct  salary_offered  \
0           Placed               7           90.00        738963.0   
1       Not Placed               8           91.67             NaN   
2           Placed               5           60.83        501904.0   
3       Not Placed               8           59.17             NaN   
4           Placed               9           84.17        970105.0   

   total_score  
0          125  
1          140  
2          122  
3          145  
4           91

The total_score column is the sum of avg_quiz_score and project_score, representing a student's overall performance in the course.


In [43]:
#13. Perform named aggregations on city to calculate total_students, placed_students, and average_salary.
calculate= df.groupby("city").agg(
    total_students=("student_id", "count"),
    placed_students=("placement_status", "sum"),
    average_salary=("salary_offered", "mean"))
calculate.head()

total_students                                    placed_students  \
city                                                                           
Bangalore              82  Not PlacedPlacedPlacedNot PlacedPlacedNot Plac...   
Chennai                73  PlacedNot PlacedPlacedNot PlacedPlacedNot Plac...   
Delhi                  68  Not PlacedNot PlacedPlacedPlacedPlacedNot Plac...   
Hyderabad             100  PlacedNot PlacedNot PlacedPlacedNot PlacedPlac...   
Mumbai                 95  PlacedPlacedPlacedPlacedPlacedNot PlacedNot Pl...   

           average_salary  
city                       
Bangalore   805241.500000  
Chennai     823148.090909  
Delhi       759655.918919  
Hyderabad   703994.405797  
Mumbai      767943.218750

in this interpritation show the average salary city wise the average salary show the 823148 city is chennai and student 73 bud hydrabad city are placement rate is very less ut  student count id very high like 100  the average placement and student count is lies on Bangalore,Delhi	and Mumbai.	

In [46]:
df["enrollment_date"] = pd.to_datetime(df["enrollment_date"])

In [48]:
#14. Extract year from enrollment_date and add it as a new column.
df["year"] =df["enrollment_date"].dt.year
df.head()

student_id student_name  gender       city enrollment_date         course  \
0        1001    Student_1    Male  Hyderabad      2024-11-25  Generative AI   
1        1002    Student_2  Female  Hyderabad      2024-09-29        AI & ML   
2        1003    Student_3    Male       Pune      2024-12-05   Data Science   
3        1004    Student_4    Male      Delhi      2024-03-09        AI & ML   
4        1005    Student_5    Male     Mumbai      2024-02-03  Generative AI   

  batch_type  total_sessions  sessions_attended  avg_quiz_score  \
0    Weekday             120                108              59   
1    Weekday             120                110              57   
2    Weekday             120                 73              49   
3    Weekend             120                 71              64   
4    Weekday             120                101              41   

   assignment_score  project_score  final_rating  learning_hours  \
0                96             66           3.7               6   
1                93             83           4.5              33   
2                72             73           3.6               8   
3                91             81           2.7               5   
4                55             50           2.1              23   

  placement_status  feedback_score  attendance_pct  salary_offered  \
0           Placed               7           90.00        738963.0   
1       Not Placed               8           91.67             NaN   
2           Placed               5           60.83        501904.0   
3       Not Placed               8           59.17             NaN   
4           Placed               9           84.17        970105.0   

   total_score  year  
0          125  2024  
1          140  2024  
2          122  2024  
3          145  2024  
4           91  2024

In [ ]:
in this interpritation add month column  and show first 4 record with there year.

In [51]:
#15. Extract month number from enrollment_date and add it as a new column
df["month"] = df["enrollment_date"].dt.month
df.head()

student_id student_name  gender       city enrollment_date         course  \
0        1001    Student_1    Male  Hyderabad      2024-11-25  Generative AI   
1        1002    Student_2  Female  Hyderabad      2024-09-29        AI & ML   
2        1003    Student_3    Male       Pune      2024-12-05   Data Science   
3        1004    Student_4    Male      Delhi      2024-03-09        AI & ML   
4        1005    Student_5    Male     Mumbai      2024-02-03  Generative AI   

  batch_type  total_sessions  sessions_attended  avg_quiz_score  ...  \
0    Weekday             120                108              59  ...   
1    Weekday             120                110              57  ...   
2    Weekday             120                 73              49  ...   
3    Weekend             120                 71              64  ...   
4    Weekday             120                101              41  ...   

   project_score  final_rating  learning_hours  placement_status  \
0             66           3.7               6            Placed   
1             83           4.5              33        Not Placed   
2             73           3.6               8            Placed   
3             81           2.7               5        Not Placed   
4             50           2.1              23            Placed   

  feedback_score  attendance_pct  salary_offered  total_score  year  month  
0              7           90.00        738963.0          125  2024     11  
1              8           91.67             NaN          140  2024      9  
2              5           60.83        501904.0          122  2024     12  
3              8           59.17             NaN          145  2024      3  
4              9           84.17        970105.0           91  2024      2  

[5 rows x 21 columns]

in this interpritation add month column  and show first 4 record with there month number.

In [54]:
#16. Extract month name from enrollment_date and add it as a new column.
df["MonthName"] = df["enrollment_date"].dt.month_name()
df.head()

student_id student_name  gender       city enrollment_date         course  \
0        1001    Student_1    Male  Hyderabad      2024-11-25  Generative AI   
1        1002    Student_2  Female  Hyderabad      2024-09-29        AI & ML   
2        1003    Student_3    Male       Pune      2024-12-05   Data Science   
3        1004    Student_4    Male      Delhi      2024-03-09        AI & ML   
4        1005    Student_5    Male     Mumbai      2024-02-03  Generative AI   

  batch_type  total_sessions  sessions_attended  avg_quiz_score  ...  \
0    Weekday             120                108              59  ...   
1    Weekday             120                110              57  ...   
2    Weekday             120                 73              49  ...   
3    Weekend             120                 71              64  ...   
4    Weekday             120                101              41  ...   

   final_rating  learning_hours  placement_status  feedback_score  \
0           3.7               6            Placed               7   
1           4.5              33        Not Placed               8   
2           3.6               8            Placed               5   
3           2.7               5        Not Placed               8   
4           2.1              23            Placed               9   

  attendance_pct  salary_offered  total_score  year  month  MonthName  
0          90.00        738963.0          125  2024     11   November  
1          91.67             NaN          140  2024      9  September  
2          60.83        501904.0          122  2024     12   December  
3          59.17             NaN          145  2024      3      March  
4          84.17        970105.0           91  2024      2   February  

[5 rows x 22 columns]

in this interpritation add month column  and show first 4 record extract with there month name.

In [57]:
#17. Extract day name from enrollment_date and add it as a new column
df["DayName"] = df["enrollment_date"].dt.day_name()
df.head()

student_id student_name  gender       city enrollment_date         course  \
0        1001    Student_1    Male  Hyderabad      2024-11-25  Generative AI   
1        1002    Student_2  Female  Hyderabad      2024-09-29        AI & ML   
2        1003    Student_3    Male       Pune      2024-12-05   Data Science   
3        1004    Student_4    Male      Delhi      2024-03-09        AI & ML   
4        1005    Student_5    Male     Mumbai      2024-02-03  Generative AI   

  batch_type  total_sessions  sessions_attended  avg_quiz_score  ...  \
0    Weekday             120                108              59  ...   
1    Weekday             120                110              57  ...   
2    Weekday             120                 73              49  ...   
3    Weekend             120                 71              64  ...   
4    Weekday             120                101              41  ...   

   learning_hours  placement_status  feedback_score  attendance_pct  \
0               6            Placed               7           90.00   
1              33        Not Placed               8           91.67   
2               8            Placed               5           60.83   
3               5        Not Placed               8           59.17   
4              23            Placed               9           84.17   

  salary_offered  total_score  year  month  MonthName   DayName  
0       738963.0          125  2024     11   November    Monday  
1            NaN          140  2024      9  September    Sunday  
2       501904.0          122  2024     12   December  Thursday  
3            NaN          145  2024      3      March  Saturday  
4       970105.0           91  2024      2   February  Saturday  

[5 rows x 23 columns]

in this interpritation add dayname column  and show first 4 record with there day

In [60]:
#18. Extract quarter from enrollment_date and add it as a new column.
df["Quarter"] = df["enrollment_date"].dt.quarter
df.head()

student_id student_name  gender       city enrollment_date         course  \
0        1001    Student_1    Male  Hyderabad      2024-11-25  Generative AI   
1        1002    Student_2  Female  Hyderabad      2024-09-29        AI & ML   
2        1003    Student_3    Male       Pune      2024-12-05   Data Science   
3        1004    Student_4    Male      Delhi      2024-03-09        AI & ML   
4        1005    Student_5    Male     Mumbai      2024-02-03  Generative AI   

  batch_type  total_sessions  sessions_attended  avg_quiz_score  ...  \
0    Weekday             120                108              59  ...   
1    Weekday             120                110              57  ...   
2    Weekday             120                 73              49  ...   
3    Weekend             120                 71              64  ...   
4    Weekday             120                101              41  ...   

   placement_status  feedback_score  attendance_pct  salary_offered  \
0            Placed               7           90.00        738963.0   
1        Not Placed               8           91.67             NaN   
2            Placed               5           60.83        501904.0   
3        Not Placed               8           59.17             NaN   
4            Placed               9           84.17        970105.0   

  total_score  year  month  MonthName   DayName  Quarter  
0         125  2024     11   November    Monday        4  
1         140  2024      9  September    Sunday        3  
2         122  2024     12   December  Thursday        4  
3         145  2024      3      March  Saturday        1  
4          91  2024      2   February  Saturday        1  

[5 rows x 24 columns]

in this interpritaion are add quarter column in first 4 data 

In [63]:
#19. Group the data by year and calculate the number of student enrollments.
df.groupby("year")["student_id"].count()

year
2024    505
Name: student_id, dtype: int64

in this intepritation there is only one year are show is 2024 year are show 505 enrollments so in this years show the highest enrollements 

In [66]:
#20. Group the data by month name and calculate average attendance percentage.
df.groupby("MonthName")["attendance_pct"].count()

MonthName
April        49
August       42
December     37
February     37
January      50
July         30
June         41
March        40
May          45
November     49
October      52
September    33
Name: attendance_pct, dtype: int64

in this interpritation are the average attendance are august, may and june 

In [70]:
#21. Group the data by quarter and calculate average total_score.
df["total_score"] = df["avg_quiz_score"] + df["project_score"]
df.groupby("Quarter")["total_score"].mean()

Quarter
1    143.157480
2    145.074074
3    146.019048
4    145.442029
Name: total_score, dtype: float64

in this interpritation are the highest quarter range are 146 and lowest range are 143 

In [73]:
#22. Group the data by day name and count the number of enrollments.
df.groupby("DayName")["student_id"].count()

DayName
Friday       74
Monday       79
Saturday     86
Sunday       74
Thursday     78
Tuesday      61
Wednesday    53
Name: student_id, dtype: int64

in this interpritation are the highest data on the above list is 78% on thursday and the lowesr is 53% on wednesday 

In [76]:
#23. Identify the month with the highest number of enrollments.
df["month"] = df["enrollment_date"].dt.month
df["month"].value_counts().head(1)


month
10    52
Name: count, dtype: int64

in this interpritation the highest number of enrollmets accrosed in the month of october and its range is around 52

In [79]:
#24. Identify the quarter with the highest average project score.
#df["Quarter"] = df["project_score"].dt.quarter
#df.groupby("Quarter")["project_score"].mean()

In [81]:
#25. Group the data by year and course and calculate average quiz score.
df.groupby(["year","course"])["avg_quiz_score"].mean()

year  course        
2024  AI & ML           69.023810
      Data Analytics    69.041958
      Data Science      69.960938
      Generative AI     70.592593
Name: avg_quiz_score, dtype: float64

in this interpritation, the higher score is for the generative AI at 70% and the year is 2024 hte same year for  AI & ML,Data Analytics and Data Science at 69%

In [84]:
#26. Group the data by month name and placement_status and count students.
df.groupby(["MonthName","placement_status"])["student_id"].mean()

MonthName  placement_status
April      Not Placed          1228.533333
           Placed              1246.382353
August     Not Placed          1198.187500
           Placed              1237.884615
December   Not Placed          1190.687500
           Placed              1297.809524
February   Not Placed          1154.000000
           Placed              1252.111111
January    Not Placed          1182.210526
           Placed              1235.774194
July       Not Placed          1190.800000
           Placed              1264.500000
June       Not Placed          1270.428571
           Placed              1312.259259
March      Not Placed          1241.000000
           Placed              1271.629630
May        Not Placed          1271.769231
           Placed              1248.156250
November   Not Placed          1277.277778
           Placed              1239.516129
October    Not Placed          1298.764706
           Placed              1271.942857
September  Not Placed     

in this interpritation, the placement status is done according to the students countsmot the also placements_status show not placed and placed.

In [87]:
#27. Group the data by city and year and calculate placement count.
df.groupby(["city","year"])["placement_status"].count()

city       year
Bangalore  2024     82
Chennai    2024     73
Delhi      2024     68
Hyderabad  2024    100
Mumbai     2024     95
Pune       2024     87
Name: placement_status, dtype: int64

in this interpritation show column wise city are here so the highest placement count is of hydrabad city and its year is 2024 and its highest placement count is 100 and second after it is mumbai with 95 placement count in the same year and also the city with the lowest placement count is delhi with 68

In [90]:
#28. Filter grouped year-wise enrollments to display only years with more than 100 enrollments.
year_count=df.groupby("year")["student_id"].count()
year_count[year_count>100]

year
2024    505
Name: student_id, dtype: int64

in this intepritation there is only one year are show  more than 100 enrollments 2024 year are show 505 enrollments so in this years show the highest enrollements 

In [93]:
#29. Filter grouped course-wise performance to show only courses with average total_score above 220.
course=df.groupby("course")["total_score"].mean()
course[course>220]

Series([], Name: total_score, dtype: float64)

in this interpritation  there is no total score above 220 in this course, so no data is printed here.

In [96]:
#30. Compare average attendance percentage across quarters.
df.groupby("Quarter")["attendance_pct"].mean()

Quarter
1    72.434252
2    72.629704
3    72.857238
4    75.525362
Name: attendance_pct, dtype: float64

in this interpritation are comparing quarters wise attendanse are showing so the quarter no.1,2,3 are same attendance showing and quarter number 4 are hightest average attendance are show.

In [99]:
#31. Calculate average salary offered per course using groupby.
df.groupby("course")["salary_offered"].mean()

course
AI & ML           785484.230769
Data Analytics    788028.545455
Data Science      781899.318182
Generative AI     713728.661972
Name: salary_offered, dtype: float64

interpritation of this data are average salary offered in this course AI & ML,Data Analytics,Data Science the same range of salary offered and generative  very less amount of salary are offered.

In [102]:
#32. Calculate average salary offered per city using groupby.
df.groupby("city")["salary_offered"].mean()

city
Bangalore    805241.500000
Chennai      823148.090909
Delhi        759655.918919
Hyderabad    703994.405797
Mumbai       767943.218750
Pune         779828.883333
Name: salary_offered, dtype: float64

interpritation of this column are the highest salary offered in chennai city 823148 and lowest salary offered are in hydedabad city 703994.

In [105]:
#33. Identify top 3 cities based on average salary offered.
df.groupby("city")["salary_offered"].mean().sort_values(ascending=False)
df.head(3)

student_id student_name  gender       city enrollment_date         course  \
0        1001    Student_1    Male  Hyderabad      2024-11-25  Generative AI   
1        1002    Student_2  Female  Hyderabad      2024-09-29        AI & ML   
2        1003    Student_3    Male       Pune      2024-12-05   Data Science   

  batch_type  total_sessions  sessions_attended  avg_quiz_score  ...  \
0    Weekday             120                108              59  ...   
1    Weekday             120                110              57  ...   
2    Weekday             120                 73              49  ...   

   placement_status  feedback_score  attendance_pct  salary_offered  \
0            Placed               7           90.00        738963.0   
1        Not Placed               8           91.67             NaN   
2            Placed               5           60.83        501904.0   

  total_score  year  month  MonthName   DayName  Quarter  
0         125  2024     11   November    Monday        4  
1         140  2024      9  September    Sunday        3  
2         122  2024     12   December  Thursday        4  

[3 rows x 24 columns]

in the table show the 3 cities based on average salary offered like hyderabad 738963 and pune 501904.0

In [122]:
#34. Identify the course with the highest placement rate using groupby analysis.
(df["placement_status"]=="Placed").groupby(df["course"]).mean().sort_values(ascending=False).head(1)

course
Data Analytics    0.692308
Name: placement_status, dtype: float64

interpritation are the highest placement rate data analytics course 

In [126]:
#35. Group the data by batch_type and quarter to analyze attendance patterns.
df.groupby(["batch_type","Quarter"])["attendance_pct"].mean().sort_values(ascending=False).head()

batch_type  Quarter
Weekend     4          77.100548
Weekday     3          74.947917
            4          73.756308
Weekend     2          73.628462
            1          73.149412
Name: attendance_pct, dtype: float64

the interpritation day by batch_type and quarter are quarter 4 and weekend batch are in the 77.10% 

In [130]:
#36. Group the data by month name and course to analyze enrollment distribution.
df.groupby(["MonthName","course"])["student_id"].count()

MonthName  course        
April      AI & ML            7
           Data Analytics    17
           Data Science      13
           Generative AI     12
August     AI & ML           12
           Data Analytics    15
           Data Science       6
           Generative AI      9
December   AI & ML            9
           Data Analytics    11
           Data Science      11
           Generative AI      6
February   AI & ML           10
           Data Analytics    10
           Data Science      11
           Generative AI      6
January    AI & ML           14
           Data Analytics    15
           Data Science      11
           Generative AI     10
July       AI & ML           10
           Data Analytics     5
           Data Science       8
           Generative AI      7
June       AI & ML            8
           Data Analytics    13
           Data Science      13
           Generative AI      7
March      AI & ML           13
           Data Analytics    11
           Dat

the interpritaion on this code are month name and course to analyze enrollment distribution is this hightest enrollment are january and the lowest enrollment are september

In [136]:
#37. Identify days of the week with maximum enrollments.
df.groupby("DayName")["student_id"].count()

DayName
Friday       74
Monday       79
Saturday     86
Sunday       74
Thursday     78
Tuesday      61
Wednesday    53
Name: student_id, dtype: int64

interpritation are the days of the week with maximum enrollments are tuesday 61%

In [138]:
#38. Analyze placement status distribution across quarters.
df.groupby(["Quarter","placement_status"])["student_id"].count()

Quarter  placement_status
1        Not Placed          42
         Placed              85
2        Not Placed          42
         Placed              93
3        Not Placed          34
         Placed              71
4        Not Placed          51
         Placed              87
Name: student_id, dtype: int64

the interpritation on this code are this result helps you understand how placement performance varies by quarter indicating which quarters have better or bad placement rates